File: k_projects/classic/merge_sort/merge_sort.go
# by Don24Kane    

# GOlang

package main

import (
	"fmt"
)

func merge(v []int, left, mid, right int) {
	n1 := (mid - left + 1)
	n2 := (right - mid)

	L := make([]int, n1)
	R := make([]int, n2)

	for i := 0; i < n1; i++ {
		L[i] = v[left+i]
	}
	for j := 0; j < n2; j++ {
		R[j] = v[mid+1+j]
	}

	i, j, k := 0, 0, left

	for i < n1 && j < n2 {
		if L[i] <= R[j] {
			v[k] = L[i]
			i++
		} else {
			v[k] = R[j]
			j++
		}
		k++
	}

	for i < n1 {
		v[k] = L[i]
		i++
		k++
	}

	for j < n2 {
		v[k] = R[j]
		j++
		k++
	}
}

func mergeSort(v []int, left, right int) {
	if left >= right {
		return
	}

	mid := left + (right-left)/2
	mergeSort(v, left, mid)
	mergeSort(v, mid+1, right)
	merge(v, left, mid, right)
}

func printSolution(v []int) {
	for _, value := range v {
		fmt.Print(value, " ")
	}
	fmt.Println()
}

func main() {
	var n int
	fmt.Print("Number of elements: ")
	fmt.Scan(&n)

	v := make([]int, n)
	fmt.Print("Elements: ")
	for i := 0; i < n; i++ {
		fmt.Scan(&v[i])
	}

	mergeSort(v, 0, n-1)

	fmt.Println("Sorted vector:")
	printSolution(v)
}

File: k_projects/classic/merge_sort/merge_sort.py
# by Don24Kane    

# Python

def merge(v, left, mid, right):
    n1 = (len(v) // 2) + 1
    n2 = len(v) // 2

    L = v[left:left + n1]
    R = v[mid + 1:mid + 1 + n2]

    i = 0
    j = 0
    k = left

    while i < n1 and j < n2:
        if L[i] <= R[j]:
            v[k] = L[i]
            i += 1
        else:
            v[k] = R[j]
            j += 1
        k += 1

    while i < n1:
        v[k] = L[i]
        i += 1
        k += 1

    while j < n2:
        v[k] = R[j]
        j += 1
        k += 1

def mergeSort(v, left, right):
    if left >= right:
        return

    mid = left + (right - left) // 2
    mergeSort(v, left, mid)
    mergeSort(v, mid + 1, right)
    merge(v, left, mid, right)

def printSolution(v):
    for i in range(len(v)):
        print(v[i], end=" ")
    print()

if __name__ == "__main__":
    v = []
    n = int(input("Number of elements: "))
    print("Elements: ", end="")
    v = [int(input()) for _ in range(n)]

    mergeSort(v, 0, n - 1)

    print("Sorted vector: ")
    printSolution(v)

